compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -1583,15 +1592,33 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let (iter_pat, iter_pat_nid) =
             self.pat_ident_binding_mode(head_span, iter, hir::BindingAnnotation::MUT);
 
-        // `match Iterator::next(&mut iter) { ... }`
         let match_expr = {
             let iter = self.expr_ident(head_span, iter, iter_pat_nid);
-            let ref_mut_iter = self.expr_mut_addr_of(head_span, iter);
-            let next_expr = self.expr_call_lang_item_fn(
-                head_span,
-                hir::LangItem::IteratorNext,
-                arena_vec![self; ref_mut_iter],
-            );
+            let next_expr = if is_await {
+                // `async_iter_next(unsafe { Pin::new_unchecked(&mut iter) }).await`
+                let future = self.expr_mut_addr_of(head_span, iter);
+                let future = self.arena.alloc(self.expr_call_lang_item_fn_mut(
+                    head_span,
+                    hir::LangItem::PinNewUnchecked,
+                    arena_vec![self; future],
+                ));
+                let future = self.expr_unsafe(future);
+                let future = self.expr_call_lang_item_fn(
+                    head_span,
+                    hir::LangItem::AsyncIterNext,
+                    arena_vec![self; future],
+                );
+                let kind = self.make_lowered_await(head_span, future);
+                self.arena.alloc(hir::Expr { hir_id: self.next_id(), kind, span: head_span })
+            } else {
+                // `match Iterator::next(&mut iter) { ... }`
```

Comment:
This comment seems wrong

`// Iterator::next(&mut iter)`

then below on line 1624 it should be:

`// match $next_expr { .. }`